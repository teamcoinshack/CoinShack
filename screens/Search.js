import React, { Component } from 'react';
import {
  Text,
  View,
  StyleSheet,
  FlatList,
  Dimensions,
  TouchableOpacity,
} from 'react-native';
import { Avatar } from 'react-native-elements';
import db from '../Database.js';
import Searchbar from '../components/Searchbar.js';

const background = '#373b48';

export default class Search extends Component {

  constructor(props) {
    super(props);

    this.state = {
      results: [],
      loading: false,
      query: '',
      callback: null,
    }
    this.fastLoad = this.fastLoad.bind(this);
    this.renderRow = this.renderRow.bind(this);
    this.load = this.load.bind(this);
  }

  load(friend) {
    this.props.navigation.navigate('FriendsProfile',{
      uid: this.state.uid,
      friendName: friend.username,
      friendUid: friend.uid,
      friendEmail: friend.email,
      callback: this.state.callback,
    })
  }

  async componentDidMount() {
    try {
      const { navigation } = this.props;
      const callback = navigation.getParam('callback', null);
      this.setState({
        callback: callback,
      });
    } catch (error) {
      console.log(error);
    }
  }

  async fastLoad(query) {
    try {
      this.setState({
        loading: true,
        query: query,
      });
      if (query === '') {
        this.setState({
          results: [],
          loading: false,
        })
        return;
      }
      let res = await db.search(query);
      res = await Promise.all(
                    res.map(async function(uid) {
                      try {
                        let obj = {};
                        const snap = await db.getData(uid);
                        const snapped = snap.val();
                        obj.uid = uid;
                        obj.username = ('username' in snapped) 
                                        ? snapped.username
                                        : 'No name :(';
                        obj.email = snapped.email;
                        obj.title_id = snapped.title_id
                        obj.image = require('../assets/icons/noPic.png');
                        return obj;
                      } catch(error) {
                        console.log(error);
                      }
                    })
                  )
      this.setState({
        results: res,
      }, () => this.setState({
          loading: false,   
        })
      );
    } catch(error) {
      console.log(error);
    }
  }

  renderRow({item}) {
    return ( 
      <TouchableOpacity
        style={styles.row}
        onPress={() => this.load(item)}
      >
        <Avatar
          rounded
          source={item.image}
          style={styles.imageStyle}
        />
        <View style={{
          flexDirection: 'column',
          alignItems: 'flex-start',
          marginLeft: 10,
        }}>
          <Text style={styles.text1}>{item.username}</Text>
          <Text style={styles.text2}>{item.email}</Text>
          <Title title_id={item.title_id} />
        </View>
      </TouchableOpacity>
    )
  }
  
  render() {
    const loading = (
      <View style={styles.loading1}>
          <MyBar
            height={65}
            width={Math.round(Dimensions.get('window').width * 0.7)}
            flexStart={true}
          />
      </View>
    )
    const searchResults = (
      <FlatList
        style={styles.flatStyle}
        data={this.state.results}
        renderItem={this.renderRow}
        keyExtractor={item => item.uid}
      />
    )
    const noResults = (
      <View style={{
        flexDirection: 'row',
        flex: 1,
        justifyContent: 'center',
        marginTop: 20,
      }}>
        <Text style={{
          color: '#7c7c7c',
          fontWeight: '500',
          fontSize: 23,
        }}>
          No results
        </Text>
      </View>
    )
    return (
      <View style={styles.container}>
        <Searchbar 
          onChangeText={query => this.fastLoad(query)}
          search={this.state.query}
          removeText={() => this.setState({ query: '', results: [] })}
        />
        {this.state.loading
          ? loading
          : this.state.results.length === 0
            ? noResults
            : searchResults}
      </View>
    )
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignSelf: "stretch",
    backgroundColor: background,
    alignItems: 'center',
  },
  flatStyle: {
    marginTop: 10,
  },
  row: {
    elevation: 1,
    borderRadius: 5,
    backgroundColor: '#515360',
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    paddingTop: 15,
    paddingBottom: 15,
    paddingLeft: 18,
    paddingRight: 16,
    marginTop: 0,
    marginBottom: 6,
    width: Math.round(Dimensions.get('window').width - 28),
  },
  imageStyle: {
    width: 80,
    height: 80,
  },
  text1: {
    fontSize: 20,
    color: '#dbdbdb',
    fontWeight: '600',
  },
  text2: {
    fontSize: 14,
    color: '#a4a9b9',
    fontWeight: '500',
    marginBottom: 5,
  },
  text3: {
    fontSize: 20,
    color: '#faed27',
  }
});
