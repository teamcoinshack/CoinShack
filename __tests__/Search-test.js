/**
 * @format
 */

import 'react-native';
import React from 'react';
import Search from '../screens/Search.js';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
const paramFunction = (param) => {
  return () => {};
}

const createTestProps = (props: Object) => ({
  navigation: {
    getParam: paramFunction
  },
  ...props
});

jest.mock('react-native-google-signin', () => {});
test('renders correctly', () => {
  let props: any;
  props = createTestProps({});
  const tree = renderer.create(<Search {...props}/>).toJSON();
  expect(tree).toMatchSnapshot();
});
