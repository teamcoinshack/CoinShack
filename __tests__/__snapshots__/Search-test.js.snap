// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
      "backgroundColor": "#373b48",
      "flex": 1,
    }
  }
  keyboardShouldPersistTaps="always"
  style={
    Object {
      "backgroundColor": "#373b48",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "marginHorizontal": 4,
          "marginTop": 5,
        }
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 2,
              "borderColor": "#515360",
              "borderRadius": 5,
              "borderWidth": 2,
              "flexDirection": "row",
              "marginVertical": 10,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginLeft": 15,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#999999",
                    "fontSize": 26,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            keyboardType="default"
            onChangeText={[Function]}
            placeholder="Search by email/username"
            placeholderTextColor="#999999"
            rejectResponderTermination={true}
            secureTextEntry={false}
            style={
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "flex": 1,
                "fontSize": 20,
                "minHeight": 40,
                "padding": 10,
                "textAlign": "left",
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginLeft": 2,
                "marginRight": 15,
              }
            }
          >
            <Text
              allowFontScaling={false}
              onPress={[Function]}
              style={
                Array [
                  Object {
                    "color": "#999999",
                    "fontSize": 20,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#7c7c7c",
            "fontSize": 23,
            "fontWeight": "500",
          }
        }
      >
        No results
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
